### PR TITLE
Update MMDLoader for the change of TGA texture horizontal origin

### PR DESCRIPTION
**Pull request by takahirox**
#8403

### CODE REVIEW

File: examples/js/loaders/MMDLoader.js [hunk 1]
```diff
@@ -2219,7 +2219,7 @@ THREE.MMDLoader.prototype.createMesh = function ( model, texturePath, onProgress
 
 						};
 
-						function detectTextureTransparency ( image, uvs ) {
+						function detectTextureTransparency ( image, uvs, flipY ) {
```
Comment (mrdoob) [inline examples/js/loaders/MMDLoader.js]: Why do you need to pass it here?

Comment (takahirox) [inline examples/js/loaders/MMDLoader.js]: To get pixcel data of image data which corresponds to uv.
If flipY is true, y should be flipped.

https://github.com/mrdoob/three.js/pull/8472/files#diff-d8334f11849f0a2342b64fee1738146fR2296

**Comment by mrdoob** [comment]
Ah. I think I understand what you mean now. You're suggesting that `TGALoader` should invert the data internally so the output is consistant and we don't need to rely on `.flipY`. Correct?

/cc @daoshengmu

**Comment by takahirox** [comment]
Yeah, it's more straightforward for me to make the output consistent.

I'm feeling weird if I see tgaTexture.flipY !== bmp/pngTexture.flipY
though they have the same Y direction on an image editor.

http://takahirox.github.io/tmp/tga_texture/textures/cat_diff.tga
http://takahirox.github.io/tmp/tga_texture/textures/cat_diff.bmp
http://takahirox.github.io/tmp/tga_texture/textures/cat_diff.png

http://takahirox.github.io/tmp/tga_texture/index.html

    var textureLoader = new THREE.TextureLoader();
    var tgaLoader = new THREE.TGALoader();

    var geometry = new THREE.PlaneGeometry( 50, 50 );

    var tgaTexture = tgaLoader.load( 'textures/cat_diff.tga' );
    var material1 = new THREE.MeshPhongMaterial( { color: 0xffffff, map: tgaTexture } );
    var mesh1 = new THREE.Mesh( geometry, material1 );
    mesh1.position.x = - 60;
    scene.add( mesh1 );

    var bmpTexture = textureLoader.load( 'textures/cat_diff.bmp' );
    var material2 = new THREE.MeshPhongMaterial( { color: 0xffffff, map: bmpTexture } );
    var mesh2 = new THREE.Mesh( geometry, material2 );
    mesh2.position.x =    0;
    scene.add( mesh2 );

    var pngTexture = textureLoader.load( 'textures/cat_diff.png' );
    var material3 = new THREE.MeshPhongMaterial( { color: 0xffffff, map: pngTexture } );
    var mesh3 = new THREE.Mesh( geometry, material3 );
    mesh3.position.x =   60;
    scene.add( mesh3 );

    // tgaTexture.flipY: false, bmpTexture.flipY: true, pngTexture.flipY: true

**Comment by mrdoob** [comment]
Yeah, makes sense to me too.

@daoshengmu do you mind tweaking the `TGALoader` so the output is always consistent?

**Comment by takahirox** [comment]
And if we wanna let TGALoader set flipY=true,
I think I can do like this

In TGALoader._parse()

    return {
        width: header.width,
        height: header.height,
        data: rgbaData,
        magFilter: THREE.NearestFilter,
        minFilter: THREE.NearestFilter,
        flipY: true // add this line
    };

In BinaryTextureLoader.load()

    // add these lines
    if ( texData.flipY !== undefined ) {

        texture.flipY = texData.flipY;

    }

**Comment by takahirox** [comment]
BTW to be honest I still can't understand why #8403 needed to make TGA texture internal data upside down.
If you cared file internal pixel data order on DataTexture.image.data,
I think we should focus on screened image.
For example, iirc BMP also has pixel order from bottom-left to right-order like TGA with TGA_ORIGIN_BL,
but texture Y direction is same as the PNG one.

**Comment by daoshengmu** [comment]
#8403 Making parse the data upside down is in order to read pixel data correctly and needn't do flipY in the browser. As far as I know, if you upload the array buffer pixel with FLIP_Y on gl context, it will do conversion in the browser. (https://dxr.mozilla.org/mozilla-central/source/dom/canvas/TexUnpackBlob.cpp#171) Open it in Firefox, it will show "WebGL: texImage2D: Uploading ArrayBuffers with FLIP_Y or PREMULTIPLY_ALPHA is slow."

In my humble opinion, if we are using binary data, we should save the work from platform. By the way, I agree with you that we should add ```texture.flipY = texData.flipY``` at BinaryTextureLoader.load(). But, in my case, flipY is false in TGALoader._parser() because I have finished the flip while parsing.

@mrdoob @WestLangley Do you think we should make flip at our texture parser or let browsers help us do this stuff?

**Comment by mrdoob** [comment]
I think the TGALoader should return an image as if it was going to be displayed in a 2d canvas. Without requiring extra flipY logic from outside.

**Comment by daoshengmu** [comment]
```
var canvas = document.getElementById('mycanvas');
var ctx = canvas.getContext("2d");
var imgData = ctx.createImageData(512, 1024); 
var loader = new THREE.TGALoader();
var texture = loader.load( 'textures/cat_diff.tga', function finish(tex, img) {

    var data = texture.image.data;
    // copy img byte-per-byte into our ImageData
   for (var i = 0, len = 512 * 1024 * 4; i < len; i++) {
	    imgData.data[i] = data[i];
   }	
   ctx.putImageData(imgData, 0, 0);
} );

var img = new Image();
img.onload = function () {
    ctx.drawImage(img, 0, 0, 512, 1024);
}
img.src = 'textures/cat_diff.png';
```

Well, we can compare the result in Canvas2D. TGA is exactly upside down if we don't do flipY from outside.

**Comment by takahirox** [comment]
Well, to simulate as if TGA file was an image file and was going to be displayed in 2D canvas,
I think the code should be, 

	var i = 0;
	for ( var y = 1024 - 1; y >= 0; y -- ) {
		for (var x = 0; x < 512; x ++ ) {
			// flip y to the screen because of TGA_ORIGIN_BL
			imgData.data[(y*512+x)*4+0] = data[i++];
			imgData.data[(y*512+x)*4+1] = data[i++];
			imgData.data[(y*512+x)*4+2] = data[i++];
			imgData.data[(y*512+x)*4+3] = data[i++];
		}
	}

Anyways, I suppose the root topic is which one TGA file is, just a binary data file or an image file.
If we think TGA file is just a binary data file, @daoshengmu 's opinion is right.
As of me, TGA file is just an image file, so its output should be consistent to other image files'.

Even if we choose the latter one,
 there might be a room to discuss for  "WebGL: texImage2D: Uploading ArrayBuffers with FLIP_Y or PREMULTIPLY_ALPHA is slow."
(Would TGALoader accept an option to switch "consistency as image data file" or "performance"?)

**Comment by mrdoob** [comment]
> As of me, TGA file is just an image file, so its output should be consistent to other image files'.

I agree.

**Comment by WestLangley** [comment]
Should the user be responsible for setting `texture.flipY` when loading a TGA file? No. It should just work.

Should the `TGALoader` set `texture.flipY = true`? No, it does not matter. The user should not care how the loader packs the data.

**Comment by daoshengmu** [comment]
As the opinions above. I think the best way is like takahirox mentioned, making the TGALoader roll-back and set ```flipY``` inside BinaryTextureLoader to be true. Then, ask WebGL context help us invert the rgba data. Although this will have performance warning, we can provide an interface like as ```function imageData(default=true)``` to give a method for users to move the inversion at the parser.

**Comment by takahirox** [comment]
I sent PR #8519 and #8520 to make TGA pixel packing revert and make TGA output be consistent to other images'.

I'm gonna close this PR without merging if you merge the above two PRs.

Regarding to the performance warning, let's make another thread.

**Pull request closed without merging**